src/main/java/uk/gov/companieshouse/api/accounts/service/impl/AbstractServiceImpl.java

```diff
@@ -0,0 +1,41 @@
+package uk.gov.companieshouse.api.accounts.service.impl;
+
+import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.GenerateEtagUtil;
+import uk.gov.companieshouse.api.accounts.model.entity.BaseEntity;
+import uk.gov.companieshouse.api.accounts.model.rest.RestObject;
+import uk.gov.companieshouse.api.accounts.service.AbstractService;
+import uk.gov.companieshouse.api.accounts.transformer.GenericTransformer;
+
+@Service
+public abstract class AbstractServiceImpl<C extends RestObject, E extends BaseEntity> implements
+        AbstractService<C, E> {
+
+    private MongoRepository mongoRepository;
+
+    private GenericTransformer<C, E> genericTransformer;
+
+    public AbstractServiceImpl(MongoRepository mongoRepository,
+            GenericTransformer<C, E> genericTransformer) {
+        this.mongoRepository = mongoRepository;
+        this.genericTransformer = genericTransformer;
+    }
+
+    @Override
+    public C save(C rest) {
+        addEtag(rest);
+        addLinks(rest);
+        addKind(rest);
+        E baseEntity = genericTransformer.transform(rest);
+        mongoRepository.save(baseEntity);
```

Comment:
Would this work if we where to add the setting of the ID? currently we allow MongoDB to set the IDs but as we are moving away from that, we will need to set them just after the transform from the rest and just before the save to the database

Reply:
Yes. An addID method would be added to the AbstractService class - then depending on if the functionality is common it would be implemented in the AbstractServiceImpl class, if it's not common functionality it would be implemented in the relevant service implementation class.